### PR TITLE
Add Fused MoE W8A8 (Int8) Support

### PR DESCRIPTION
**Pull request by qingquansong**
Add Fused MoE W8A8 (Int8) Support since without having it int8 moe is much slower on A100 compared to bf16 with fused moe.

 [Need more tuning to make it faster]
With some minor tuning, current kernel is for 4k context length with TP=8 on 8 A100 with  max_num_batched_tokens= 64 * 4096 and 64 batch size is ~13400 toks/s for prefill stage (only slightly faster than bf16 with fused moe 12900 toks/s) and output ~23 toks/s which is much slower than bf16 output: 48.53 toks/s. Need further tuning to make it work better. A probably non-fair but still worthy comparison is that since 8*22b cannot (or at the limit) to serve with 4GPUs but the quantized one can be easily served with 4GPUs. If we serve 2 with 4GPUs on 1 node, each will have ~8500 toks/s for prefill and 24 toks/s for output, so can match the output while have 17000 toks/s per node throughput. Probably TP/EP implementation is the bottleneck that blocking the MOE to work well and Llama 70 without MoE can have 20%+ improvement even using TP8. 

The quantization strategy is posted here: https://github.com/vllm-project/llm-compressor/issues/35#issuecomment-2249709368 with channel wise 8bit quantization for weights and per tensor static input activation quantization. Expert routing gate layer is not quantized (as cutlass kernel by default only support   tensor_row % 16 = 0 input and 8 experts without doing padding does not support this. 

One more thing is the chunk strategy [here](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/fused_moe.py#L453-L490) for solving https://github.com/vllm-project/vllm/issues/5938 seems not quite useful as removing it seems also works as expected and can give slightly faster speed for both normal bf16 and int8. Maybe can remove it?

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by qingquansong** [comment]
cc @robertgshaw2-neuralmagic feel free to provide any comments or link to any POCs. Thank you!

File: vllm/model_executor/models/mixtral_quant.py [hunk 1]
```diff
@@ -48,7 +48,7 @@
 from vllm.model_executor.model_loader.weight_utils import default_weight_loader
 from vllm.model_executor.sampling_metadata import SamplingMetadata
 from vllm.sequence import IntermediateTensors, SamplerOutput
-
+from vllm.model_executor.layers.fused_moe import FusedMoE
```
[the remark this answers is not in the document]
Comment (qingquansong) [reply vllm/model_executor/models/mixtral_quant.py:51]: Sure. That makes a lot of sense. I'll try to allocate some time in recent 1-2 weeks to rebase and update.

[the remark this answers is not in the document]
Comment (qingquansong) [reply vllm/model_executor/models/mixtral_quant.py]: Hey @mgoin since I'm also using the compressed tensor format form neural magic to implement this and I saw you already have one reverted pr https://github.com/vllm-project/vllm/pull/7764/files#diff-db60d9d797517c6bdf2ce051b532fe6fa6bceab04516c6d8c2d80a2ae803abe9 that contains the compressed tensor format checking and loading. Do you suggest me to wait for the one to be checkin in again with marlin kernels or you have some better suggestions to help checkin this pr? I think one way for me is to check in the kernel part only in this pr. Or I can put merge together into the ExpertInt8 class to have another flag for w8a8 https://github.com/vllm-project/vllm/pull/7415/files#diff-312c4aa092501cccba5cce56647f4d812f843165ab0e2e8d1c4fd449d6ed60a8R125 . But this an extra config passing from use to know whether they wanna use w8a8 or w8a16, which is not good. I'm also not sure what specific quantization method that the expert_int8 pr is using which i'm assuming not GPTQ/AWQ as in compressed tensor format.  

I can also wait for you pr to be merged again with marlin kernels because otherwise I'll need to put  some extra checking here https://github.com/vllm-project/vllm/pull/7764/files#diff-e6cba5598f37013ace29f382dd3353f55275bd7137fb30a11e21155d73db9400L26 to only keep the case for w8a8 + compressed tensor format. Thank you!

[the remark this answers is not in the document]
Comment (qingquansong) [reply vllm/model_executor/models/mixtral_quant.py]: @mgoin I made a modification and if you think this solution is good as a temporary solution (before your more general version of code + marlin checked in), please feel free to provide comments and I can add some testing as well before checking in. Thank you!

[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/model_executor/models/mixtral_quant.py]: Sorry for missing these comments @qingquansong ! The Marlin MoE PR should be in now so I think it should be good for you to build on. For now I am willing to live with separate w8a8 and w8a16 configs in the triton kernel to get this feature landed, but we should replace this with at least an enum so we can reduce it to a single argument soon.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 2]
```diff
@@ -116,6 +118,13 @@ def fused_moe_kernel(
     off_experts = tl.load(expert_ids_ptr + pid_m)
     b_ptrs = b_ptr + off_experts * stride_be + (offs_k[:, None] * stride_bk +
                                                 offs_bn[None, :] * stride_bn)
+
+    if use_int8_w8a8:  # only support static per expert input activation whole tensor quantization now
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:122]: Unfortunately I don't think per-tensor scales for activations will be accurate enough for int8. Do you think dynamic per-token quantization could be implemented?

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 3]
```diff
@@ -276,19 +292,20 @@ def invoke_fused_moe_kernel(A: torch.Tensor, B: torch.Tensor, C: torch.Tensor,
...
         use_fp8_w8a8=use_fp8_w8a8,
+        use_int8_w8a8=use_int8_w8a8,
         use_int8_w8a16=use_int8_w8a16,
         **config,
     )
 
 
 def get_config_file_name(E: int, N: int, dtype: Optional[str]) -> str:
-    device_name = current_platform.get_device_name().replace(" ", "_")
+    device_name = torch.cuda.get_device_name().replace(" ", "_")
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:308]: What was the issue with using current_platform? We should really try to keep using that interface as we support other accelerators

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 4]
```diff
@@ -129,6 +132,141 @@ def forward_tpu(self,
...
+        if self.is_static_input_scheme:
+            scale = torch.nn.Parameter(torch.ones(num_experts,
+                                                  dtype=torch.float32),
+                                       requires_grad=False)
+            set_weight_attrs(scale, {
+                "needs_scalar_to_array": True,
+                **layer_kwargs
+            })
+            set_weight_attrs(scale, {"is_int8_input_scale": True})
+            layer.register_parameter("w13_input_scale", scale)
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:193]: This seems to assume per-tensor scale? We need to check the strategy to see if we can support the scheme in the checkpoint

**Review by mgoin — COMMENTED** [review]
I think we don't have many INT8 models with per-tensor scales because the accuracy drop is too severe. We generally target static per-channel scales for the weights and dynamic per-token scales for the activations for INT8 https://github.com/neuralmagic/compressed-tensors/blob/86211a6b6f30dae9cfa7ee48b0994fed261902a2/src/compressed_tensors/quantization/quant_scheme.py#L115-L130 so I would like to see that implemented. FP8 is able to get away with static per-tensor scales for weights/activations much more easily.
cc @dsikka since the weight loading could also be simplified using the recent refactor

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 2 above)
Comment (qingquansong) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:122]: yeah, that could be done but I'm a bit confused on the llm-compressor recipe setup, do you have an example recipe yml file to correctly set it (without using the class to config it in code)?

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 3 above)
Comment (qingquansong) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:308]: oh, sure, will revert it, it's just for some local testing purpose

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 4 above)
Comment (qingquansong) [reply vllm/model_executor/layers/fused_moe/layer.py:193]: yes, input is per tensor (each expert has it's own input scale)

File: vllm/model_executor/models/mixtral_quant.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qingquansong) [reply vllm/model_executor/models/mixtral_quant.py]: Thanks for the response! Do you suggest to remove this class https://github.com/vllm-project/vllm/pull/6978/files#diff-eddafffeb6f159f8c75f635d18a502fcfbf662a562b1ae7a8683a9790161a10bR135 if the llm compressor layer loading/setup schema has been fully checked in? 🤔
